icu/icu4c/source/common/uprefs.cpp

```diff
@@ -0,0 +1,579 @@
+            }
+
+            if (hourFormat[i] == L'h')
+            {
+                return CharString("h12", *status);
+            }
+        }
+    }
+    // default to a 24 hour clock as that's more common worldwide
+    return CharString("h23", *status);
+}
+
+UErrorCode getUErrorCodeFromLastError()
+{
+    DWORD error = GetLastError();
+    switch(error)
+    {
+        case ERROR_INSUFFICIENT_BUFFER:
+            return U_BUFFER_OVERFLOW_ERROR;
+
+        case ERROR_INVALID_FLAGS:
+        case ERROR_INVALID_PARAMETER:
+            return U_ILLEGAL_ARGUMENT_ERROR;
+        case ERROR_OUTOFMEMORY:
+            return U_MEMORY_ALLOCATION_ERROR;
+        default:    
+            return U_INTERNAL_PROGRAM_ERROR;
+    }
+}
+
+int32_t GetLocaleInfoExWrapper(LPCWSTR lpLocaleName, LCTYPE LCType, LPWSTR lpLCData, int cchData, UErrorCode* errorCode)
+{
+    if (U_FAILURE(*errorCode))
+    {
+        return 0;
+    }
+#ifndef UPREFS_TEST
+    *errorCode = U_ZERO_ERROR;
+    int32_t result = GetLocaleInfoEx(lpLocaleName, LCType, lpLCData, cchData);
+
+    if (result == 0)
+    {
+        *errorCode = getUErrorCodeFromLastError();
+    }
+    return result;
+#else
+    #include "uprefstest.h"
+    UPrefsTest prefTests;
+    return prefTests.MockGetLocaleInfoEx(lpLocaleName, LCType, lpLCData, cchData, errorCode);
+#endif
+}
+
+// Copies a string to a buffer if its size allows it and returns the size.
+// The returned needed buffer size includes the terminating \0 null character.
+// If the buffer's size is set to 0, the needed buffer size is returned before copying the string.
+int32_t checkBufferCapacityAndCopy(const char* uprefsString, char* uprefsBuffer, int32_t bufferSize, UErrorCode* status)
+{
+    size_t neededBufferSize = uprv_strlen(uprefsString) + 1;
+
+    RETURN_VALUE_IF(bufferSize == 0, static_cast<int32_t>(neededBufferSize));
+    RETURN_FAILURE_WITH_STATUS_IF(neededBufferSize > bufferSize, U_BUFFER_OVERFLOW_ERROR);
+
+    uprv_strcpy(uprefsBuffer, uprefsString);
+
+    return static_cast<int32_t>(neededBufferSize);
+}
+
+CharString getLocaleBCP47Tag_impl(UErrorCode* status)
+{
+    // First part of a bcp47 tag looks like an NLS user locale, so we get the NLS user locale.
+    int32_t neededBufferSize = GetLocaleInfoExWrapper(LOCALE_NAME_USER_DEFAULT, LOCALE_SNAME, nullptr, 0, status);
+
+    if (U_FAILURE(*status))
+    {
+        return CharString();
+    }
+
+    MaybeStackArray<wchar_t, 40> NLSLocale(neededBufferSize, *status);
+    RETURN_WITH_ALLOCATION_ERROR_IF(U_FAILURE(*status), status);
+    
+    int32_t result = GetLocaleInfoExWrapper(LOCALE_NAME_USER_DEFAULT, LOCALE_SNAME, NLSLocale.getAlias(), neededBufferSize, status);
+
+    if (U_FAILURE(*status))
+    {
+        return CharString();
+    }
+
+    // The NLS locale may include a non-default sort, such as de-DE_phoneb. We only want the locale name before the _.
+    wchar_t * position = wcsstr(NLSLocale.getAlias(), L"_");
```

Comment:
I think this could use `wcschr` instead, as we're not really looking for a string inside a string -- we're only looking for a single character. 

In other words:
```suggestion
    wchar_t * position = wcschr(NLSLocale.getAlias(), L'_');
```